src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemorySegment.java

```diff
@@ -849,25 +873,106 @@ static MemorySegment ofNativeRestricted() {
     int CLOSE = WRITE << 1;
 
     /**
-     * Share access mode; this segment support sharing with threads other than the owner thread (see {@link #withOwnerThread(Thread)}}).
-     * (see {@link #spliterator(MemorySegment, SequenceLayout)}).
+     * Handoff access mode; this segment support temporal bound changes (see {@link #handoff(NativeScope)} and
+     * {@link #handoff(HandoffTransform)}).
      * @see MemorySegment#accessModes()
      * @see MemorySegment#withAccessModes(int)
      */
-    int SHARE = CLOSE << 1;
+    int HANDOFF = CLOSE << 1;
 
     /**
-     * Handoff access mode; this segment support serial thread-confinement via thread ownership changes
-     * (see {@link #withOwnerThread(Thread)}).
+     * Default access mode; this is a union of all the access modes supported by memory segments.
      * @see MemorySegment#accessModes()
      * @see MemorySegment#withAccessModes(int)
      */
-    int HANDOFF = SHARE << 1;
+    int ALL_ACCESS = READ | WRITE | CLOSE | HANDOFF;
 
     /**
-     * Default access mode; this is a union of all the access modes supported by memory segments.
-     * @see MemorySegment#accessModes()
-     * @see MemorySegment#withAccessModes(int)
+     * An <em>handoff transform</em> defines a transformation which applies to an existing memory segment (the
+     * <em>source</em> segment) and results in a new segment (the <em>target</em> segment) with fresh temporal bounds
+     * featuring new characteristics, such as a different confinement thread, or a custom cleanup action, or implicit
+     * deallocation provided by a custom {@link Cleaner} instance. Handoff transforms are applied to segments using the
+     * {@link #handoff(HandoffTransform)} method.
+     * <p>
+     * There are two kinds of handoff transforms: confined (see {@link #ofConfined()}, {@link #ofConfined(Thread)}) and
+     * shared (see {@link #ofShared()}), which produce confined and shared target segments, respectively.
+     *
+     * @apiNote In the future, if the Java language permits, {@link HandoffTransform}
+     * may become a {@code sealed} interface, which would prohibit subclassing except by
+     * explicitly permitted subtypes.
+     *
+     * @implSpec
+     * Implementations of this interface are not thread-safe.
+     *
+     * @see MemorySegment#handoff(HandoffTransform)
      */
-    int ALL_ACCESS = READ | WRITE | CLOSE | SHARE | HANDOFF;
+    interface HandoffTransform {
```

Comment:
There might be a use-case for starting out with the existing confinement properties of a segment, say if just registering a cleaner or adding an attachment e.g. `HandoffTransform.ofSegment`, which is not a good name since it would imply the acceptance of a segment parameter, whereas it would be applied to the segment that accepts the handoff transform. That way it's not necessary to query a segment before constructing the correct transform.

I suspect there will only be one implementation of `HandoffTransform`, since it's really just a dumb holder of accumulated state, so it could be made into a `final class` (value-based) with a non-public constructor.

Reply:
`HandoffTransform.identity()` ?

Reply:
Yes - this is a dumb holder for accumulated state - my question in the review header was as to whether we're fine with it being dumb and mutable and thread unsafe, since that is reflected in the javadoc

Reply:
`HandoffTransform.of()`, or `ofSource`, seems clear, such that `segment.handOff(HandoffTransform.of())` returns a new segment with exactly the same properties as the source.

I think being a dumb holder of mutable state is fine. It's simple and there is no binding/reuse issue as, say, with a builder scoped to a method call.